### ISSUE TITLE
getTFeatureMT implementation

### ISSUE THREAD

**Issue by rlnx** [francktcheng/daal#4]
@francktcheng, I’m looking at the code in your fork, I wonder, do you really need to call `memcpy` within `getTFeatureMT`?
https://github.com/francktcheng/daal/blob/9115b9aa0f385590980ce24080403f2e31104f56/lang_service/java/com/intel/daal/data_management/data/java_harp_numeric_table.h#L532

It seems, you can simply create ByteBuffer at every iteration:
```C++
if( !(block[feature_idx]->resizeBuffer( 1, nrows )) ) { return; }
buf = block[feature_idx]->getBlockPtr();
jobject jbuf = (local_tls.jenv)->NewDirectByteBuffer(buf, bufferSize);
```

**Comment by rlnx** [comment]
> I tested your suggestion and compare the performance with mine. I moved the `NewDirectByteBuffer` invocation out of the iteration loop on feature length because I was not sure the new created Java object buffer at each iteration would  be recycled instantaneously by the JVM. The performance shows that your suggestion may increase the peak memory utilization at some extent. In my tests on Hugewiki, Dim=500, it slightly increases by 1.9 GB at each node (2% of the total ). 

Is it possible to get rid of multiple block descriptors `BlockDescriptor<float>** block`? At every iteration `block[feature_idx]->resizeBuffer( 1, nrows )` causes memory allocation. If we replace it with single `BlockDescriptor`, we will be able to allocate a single chunk of memory before the loop.

**Comment by francktcheng** [comment]
I think the loop iteration on BlockDescriptor<float>** block will not create redundant memory allocation because each iteration corresponds to a feature dimension.  If the HarpNumericTable has a homogeneous data layout (double, float, or int), it is ok to use a single BlockDescriptor, we may try this. For SOANumericTable, we need separated BlockDescriptors for each feature dimension.

**Comment by rlnx** [comment]
> I think the loop iteration on BlockDescriptor** block will not create redundant memory allocation because each iteration corresponds to a feature dimension. 

It doesn't create redundant memory allocation, but, as a rule, multiple memory allocation is slower than allocation of a single "big" memory chunk. Moreover, you mentioned, that you are going to call this method inside a parallel loop, so `malloc` inside parallel section may cause a lock.

**Comment by francktcheng** [comment]
@RuslanIsrafilov, yes it is true that a single memory allocation is faster than a number of malloc operations for each column. However, we may need a table that can deallocate a part of its columns in runtime ( shrink of feature data in applications like LDA ). Therefore, it is more like a SOA table instead of a homogeneous table with a single memory chunk. You mentioned a "malloc" inside parallel section may cause a lock, so we need something like scalable_allocator<T> ?

**Comment by rlnx** [comment]
Yes, in this case `scalable_allocator` is preferable, need to think how to let user select allocator in BlockDescriptor.
